cpp/include/raft/neighbors/ivf_pq_types.hpp

```diff
@@ -105,6 +105,12 @@ struct index_params : ann::index_params {
    * flag to `true` if you prefer to use as little GPU memory for the database as possible.
    */
   bool conservative_memory_allocation = false;
+  /** 
+   * The fraction of data to use during PQ codebook generation on top of the subsampled data 
+   * controlled by kmeans_trainset_fraction. The parameter is only used when PQ codebook generation 
+   * kind is PER_SUBSPACE and ignored when PQ codebook generation kind is PER_CLUSTER. 
+  */
+  double pq_codebook_trainset_fraction = 1;
```

Comment:
@cjnolet and @achirkin:
This parameter to controls how much are the training vectors subsampled while we train the PQ codebooks. In practice we have two level subsampling 
```c++
n_rows;                                                      // number of input vectors
n_train = n_rows * kmeans_trainset_fraction;                 // vectors used to train IVF clusters
n_train_codebook = n_train * pq_codebook_trainset_fraction;  // vectors used for training codebooks
```
**Question**
Is this the right way to control the number of points, or shall we have something similar to FAISS where the `max_points_per_centroids` are set?

**Background**
Codebook training is a k-means clustering operation in the PQ subspaces. The number of clusters is 2^pq_bit.

In FAISS `IndexPQ`, we have a [ProductQuantizer object](https://github.com/facebookresearch/faiss/blob/c3aa526ec7410d7ad2c8b243439ecbbc8c8c76c5/faiss/IndexPQ.h#L26) that is responsible for the PQ codebooks. Its `train` method uses a [Clustering object](https://github.com/facebookresearch/faiss/blob/c3aa526ec7410d7ad2c8b243439ecbbc8c8c76c5/faiss/impl/ProductQuantizer.cpp#L145) to run k-means clustering (this is the same class is used for trainig IVF clusters). 

Clustering defines the following [parameter to control subsampling](https://github.com/facebookresearch/faiss/blob/c3aa526ec7410d7ad2c8b243439ecbbc8c8c76c5/faiss/Clustering.h#L44-L45):
```c++
/// to limit size of dataset, otherwise the training set is subsampled
int max_points_per_centroid = 256;
```

and this parmeter is [used here](https://github.com/facebookresearch/faiss/blob/c3aa526ec7410d7ad2c8b243439ecbbc8c8c76c5/faiss/Clustering.cpp#L302).

**Notes**
- FAISS does not seem to provide an easy to access user knob for controlling the subsampling rate of codebooks.
- The number of training vectors in FAISS depends on `pq_bits` and the constant subsampling ratio.
```python
n_train_codebook = max_points_per_centroids * n_clusters
                 = max_points_per_centroids * 2**pq_bits
```

**RAFT notes**
- The idea of subsampling vectors for codebook training is in agreement with what FAISS does.
- We already do [subsampling  in RAFT](https://github.com/rapidsai/raft/blob/0c75bb1985c1fe9177cf18e9c42d071d7115af7b/cpp/include/raft/neighbors/detail/ivf_pq_build.cuh#L522-L524) when `codebook_kind`  is `per_cluster`. In that case we just set `n_train_codebook = 256*max(2**pq_bits, pq_dim)`.

Reply:
That's indeed a hard question. On the one hand, having this parameter defined as a fraction nicely mirrors the existing parameter `kmeans_trainset_fraction`. On the other hand, an optimal value for such parameter is not a constant; it clearly depends on the size of the train set and on the codebook size (2^pq_bits). The `max_points_per_centroid` or something like `max_train_points_per_pq_code` on its own makes much more sense to me.
Perhaps, I would favor slightly more the latter, max-points parameter.

Reply:
How does FAISS control the fraction of the first-level subsampling (subsampling for input table)? In RAFT we have `kmeans_trainset_fraction`, but my understanding is FAISS samples `max_points_per_centroid` elements per `n_list`. If this is correct, in order to match with FAISS, do we want to remove the existing `kmeans_trainset_fraction` parameter as well?

Reply:
I agree with @achirkin, that a max-points parameter would make sense.

@abc99lr , your understanding is correct. But this PR does not need to change  `kmeans_trainset_fration` which is used for the first-level subsampling. We are considering to introduce a second parameter for the pq codebook, For now it is enough to align that parameter with FAISS.

Reply:
Thanks for the clarification. I'll modify this PR. When would https://github.com/rapidsai/raft/pull/2077 be merged? I'd like to push and test the change after https://github.com/rapidsai/raft/pull/2077 is merged.